File: libs/statemanager/src/main/kotlin/mono/state/command/mouse/AddLineMouseCommand.kt
/*
 * Copyright (c) 2023, tuanchauict
 */

package mono.state.command.mouse

import kotlin.math.abs
import mono.common.MouseCursor
import mono.common.exhaustive
import mono.graphics.geo.DirectedPoint
import mono.graphics.geo.MousePointer
import mono.graphics.geo.Point
import mono.shape.command.MoveLineAnchor
import mono.shape.connector.ShapeConnectorUseCase
import mono.shape.selection.SelectedShapeManager
import mono.shape.shape.Line
import mono.state.command.CommandEnvironment

/**
 * A [MouseCommand] for adding a new Line shape.
 */
internal class AddLineMouseCommand : MouseCommand {
    override val mouseCursor: MouseCursor = MouseCursor.CROSSHAIR

    private var workingShape: Line? = null

    private val hoverShapeManager = HoverShapeManager.forLineConnectHover()

    override fun execute(
        environment: CommandEnvironment,
        mousePointer: MousePointer
    ): MouseCommand.CommandResultType =
        when (mousePointer) {
            is MousePointer.Down -> {
                workingShape = environment.createLineAndAdjustStartAnchor(mousePointer)
                environment.clearSelectedShapes()
                MouseCommand.CommandResultType.WORKING
            }

            is MousePointer.Drag -> {
                environment.changeEndAnchor(
                    environment,
                    mousePointer.boardCoordinate,
                    mousePointer.isWithShiftKey,
                    isUpdateConfirmed = false
                )
                MouseCommand.CommandResultType.WORKING
            }

            is MousePointer.Up -> {
                environment.changeEndAnchor(
                    environment,
                    mousePointer.boardCoordinate,
                    mousePointer.isWithShiftKey,
                    isUpdateConfirmed = true
                )
                environment.addSelectedShape(workingShape)
                MouseCommand.CommandResultType.DONE
            }

            is MousePointer.Move,
            is MousePointer.Click,
            is MousePointer.DoubleClick,
            MousePointer.Idle -> MouseCommand.CommandResultType.UNKNOWN
        }.exhaustive

    private fun CommandEnvironment.createLineAndAdjustStartAnchor(
        mousePointer: MousePointer.Down
    ): Line {
        val edgeDirection = getEdgeDirection(mousePointer.boardCoordinate)
        val direction = edgeDirection?.normalizedDirection ?: DirectedPoint.Direction.HORIZONTAL
        val line = Line(
            DirectedPoint(direction, mousePointer.boardCoordinate),
            DirectedPoint(DirectedPoint.Direction.VERTICAL, mousePointer.boardCoordinate),
            parentId = workingParentGroup.id
        )
        addShape(line)

        val connectShape = ShapeConnectorUseCase.getConnectableShape(
            line.startPoint,
            getShapes(mousePointer.boardCoordinate)
        )
        if (connectShape != null) {
            shapeManager.shapeConnector.addConnector(line, Line.Anchor.START, connectShape)
        }
        return line
    }

    private fun CommandEnvironment.changeEndAnchor(
        environment: CommandEnvironment,
        point: Point,
        isStraightLineMode: Boolean,
        isUpdateConfirmed: Boolean
    ) {
        val line = workingShape ?: return
        val endPoint = adjustEndPoint(line.startPoint.point, point, isStraightLineMode)
        val edgeDirection = environment.getEdgeDirection(point)
        val direction = edgeDirection?.normalizedDirection ?: line.getDirection(Line.Anchor.END)
        val anchorPointUpdate = Line.AnchorPointUpdate(
            Line.Anchor.END,
            DirectedPoint(direction, endPoint)
        )

        val connectShape = hoverShapeManager.getHoverShape(environment, anchorPointUpdate.point)
        environment.setFocusingShape(
            connectShape.takeIf { !isUpdateConfirmed },
            SelectedShapeManager.ShapeFocusType.LINE_CONNECTING
        )
        shapeManager.execute(
            MoveLineAnchor(
                line,
                anchorPointUpdate,
                isUpdateConfirmed,
                justMoveAnchor = false,
                connectShape = connectShape
            )
        )
    }

    private fun adjustEndPoint(
        startPoint: Point,
        candidateEndPoint: Point,
        isStraightLineMode: Boolean
    ): Point {
        if (!isStraightLineMode) {
            return candidateEndPoint
        }
        val width = abs(startPoint.left - candidateEndPoint.left)
        val height = abs(startPoint.top - candidateEndPoint.top)
        return if (width > height) {
            Point(candidateEndPoint.left, startPoint.top)
        } else {
            Point(startPoint.left, candidateEndPoint.top)
        }
    }
}

File: monosketch-svelte/src/lib/mono/state-manager/command/mouse/add-line-mouse-command.ts
import { DirectedPoint, Direction, Point } from "$libs/graphics-geo/point";
import { MoveLineAnchor } from "$mono/shape/command/line-commands";
import { ShapeConnectorUseCase } from "$mono/shape/connector/line-connector";
import { ShapeFocusType } from "$mono/shape/selected-shape-manager";
import { Line } from "$mono/shape/shape/line";
import { LineAnchor } from "$mono/shape/shape/linehelper";
import type { CommandEnvironment } from "$mono/state-manager/command-environment";
import { HoverShapeManager } from "$mono/state-manager/command/mouse/hover-shape-manager";
import { MouseCommand } from "$mono/state-manager/command/mouse/mouse-command";
import { MouseCursor } from "$mono/workspace/mouse/cursor-type";
import { type MousePointer, type MousePointerDown, MousePointerType } from "$mono/workspace/mouse/mouse-pointer";
import CommandResultType = MouseCommand.CommandResultType;

/**
 * A [MouseCommand] for adding a new Line shape.
 */
export class AddLineMouseCommand implements MouseCommand {
    mouseCursor: MouseCursor = MouseCursor.CROSSHAIR;

    private workingShape: Line | null = null;
    private readonly hoverShapeManager: HoverShapeManager = HoverShapeManager.forLineConnectHover();

    execute(environment: CommandEnvironment, mousePointer: MousePointer): MouseCommand.CommandResultType {
        switch (mousePointer.type) {
            case MousePointerType.DOWN: {
                this.workingShape = this.createLineAndAdjustStartAnchor(environment, mousePointer);
                environment.clearSelectedShapes();
                return CommandResultType.WORKING;
            }
            case MousePointerType.DRAG: {
                this.changeEndAnchor(environment, mousePointer.boardCoordinate, mousePointer.isWithShiftKey, false);
                return CommandResultType.WORKING;
            }
            case MousePointerType.UP: {
                this.changeEndAnchor(environment, mousePointer.boardCoordinate, mousePointer.isWithShiftKey, true);
                environment.addSelectedShape(this.workingShape);
                return CommandResultType.DONE;
            }
            case MousePointerType.MOVE:
            case MousePointerType.CLICK:
            case MousePointerType.DOUBLE_CLICK:
            case MousePointerType.IDLE:
                return CommandResultType.UNKNOWN;
        }
    }

    private createLineAndAdjustStartAnchor(
        environment: CommandEnvironment,
        mousePointer: MousePointerDown,
    ): Line {
        const edgeDirection = environment.getEdgeDirection(mousePointer.boardCoordinate);
        const direction = edgeDirection !== null ? Direction.normalizedDirection(edgeDirection) : Direction.HORIZONTAL;

        const line = Line.fromPoints(
            {
                startPoint: new DirectedPoint(direction, mousePointer.boardCoordinate),
                endPoint: new DirectedPoint(Direction.VERTICAL, mousePointer.boardCoordinate),
                parentId: environment.workingParentGroup.id,
            },
        );

        environment.addShape(line);

        const connectShape = ShapeConnectorUseCase.getConnectableShape(
            line.startPoint.point,
            environment.getShapes(mousePointer.boardCoordinate),
        );

        if (connectShape !== null) {
            environment.shapeManager.shapeConnector.addConnector(line, LineAnchor.START, connectShape);
        }
        return line;
    }

    private changeEndAnchor(
        environment: CommandEnvironment,
        point: Point,
        isStraightLineMode: boolean,
        isUpdateConfirmed: boolean,
    ): void {
        const line = this.workingShape;
        if (!line) return;

        const endPoint = this.adjustEndPoint(line.startPoint.point, point, isStraightLineMode);
        const edgeDirection = environment.getEdgeDirection(point);
        const direction =
            edgeDirection !== null ? Direction.normalizedDirection(edgeDirection) : line.getDirection(LineAnchor.END);

        const anchorPointUpdate = {
            anchor: LineAnchor.END,
            point: new DirectedPoint(direction, endPoint),
        };

        const connectShape = this.hoverShapeManager.getHoverShape(
            environment,
            anchorPointUpdate.point.point,
        );

        environment.setFocusingShape(
            isUpdateConfirmed ? null : connectShape,
            ShapeFocusType.LINE_CONNECTING,
        );

        environment.shapeManager.execute(
            new MoveLineAnchor(
                line,
                anchorPointUpdate,
                isUpdateConfirmed,
                false, // justMoveAnchor
                connectShape,
            ),
        );
    }

    private adjustEndPoint(
        startPoint: Point,
        candidateEndPoint: Point,
        isStraightLineMode: boolean,
    ): Point {
        if (!isStraightLineMode) {
            return candidateEndPoint;
        }

        const width = Math.abs(startPoint.left - candidateEndPoint.left);
        const height = Math.abs(startPoint.top - candidateEndPoint.top);

        if (width > height) {
            return new Point(candidateEndPoint.left, startPoint.top);
        } else {
            return new Point(startPoint.left, candidateEndPoint.top);
        }
    }
}
